datagovsg/send
# Add logout functionality and button on header

## Description

[[VT-144] WEB-L-01: No Logout Button Present on “Send” Page](https://team-1582686133065.atlassian.net/browse/VT-144?atlOrigin=eyJpIjoiOTJmNjM1ZjIxNmRkNDdlZmEyOTI2MWFhMWY0MWE3ZTYiLCJwIjoiaiJ9)

**Problem:**  
During the testing of this application, consultant observed that “Send” web page of the application do not have a logout button. Without a logout function, a user can close the browser but has no easy way of terminating the server-side session.

**Changelog:**  

- Backend logic: Add a `/api/auth/logout` route
  - Updated `server/routes/index.js` to reference a new handler `server/routes/logout.js`
- Frontend logic:
  - A link in `app/ui/header.js` calls to the frontend route `/vault-logout`
  - The route handler for `/vault-logout` in `app/routes.js` emits an event `vaultLogout`
  - The event handler `vaultLogout` in `app/controller.js` calls `invalidateVaultSession()` from `app/api.js`
  - The method `app/api.js` does a GET fetch request from the backend server route `/api/auth/logout`
- Frontend UI:
   - Add logout button
   - Add back button and link it up with the `LOGIN_URL` environment variable
- Misc:
  - Add `npm run dev` script, update `README.md` to reference this new script.
  - Refactor `VAULT_FRONTEND_URL` as `LOGIN_URL`

## Decisions/Tradeoffs
**Suggested solutions:**

- [x] Wire up a logout button (Less effort)
- [ ] Duplicate the existing Vault header and remove the search so users can still access other parts of Vault (Preferred for UX experience)

**Decision:**  
The first option was chosen in consideration of:

- Code complexity: have to replicate business logic for main Vault app header here too (e.g. only show admin panel for admin users)
- Code maintainability:
  - future UI changes/improvements to 1 header will require an equivalent update to the other, which is not so straightforward given the main Vault project and this builds html pages differently, and this only uses `css` files unlike the main project which uses `scss` files
  - update to business logic for main Vault app will need to be replicated here too

## Screenshots (WIP)
![image](https://user-images.githubusercontent.com/13582874/82418087-1f1a6800-9aaf-11ea-94c2-112a7efb88c1.png)

## How to test
1. Visit the Vault Send page
2. Click the Back button > It should return you to the main Vault application.
3. Click the Logout button > It should log you out and redirect you to the login page on the main Vault application.

## Type of change

- New feature (non-breaking change which adds functionality)
